### PR TITLE
Extend the Solidity AST for precise locations of names

### PR DESCRIPTION
**Pull request by christianparpart**
This is an extracted commit of PR https://github.com/ethereum/solidity/pull/8125

The actual SourceLocation on an ASTNode is representing the whole
ASTNode whereas in an LSP context (for example) you are also interested in the
SourceLocation of a name of that construct (for example variable declaration, function definition, ...).

This PR is a split-out of the LSP PR that only covers modifying the AST.

With regards to tests: I think this would be implicitly tests within the `solls` (LSP server) test cases that will expect the correct source locations for symbols in tests for e.g. "find all references", "semantic highlighting", "go to definition".

### CODE REVIEW

File: libsolidity/parsing/Parser.cpp [hunk 1]
```diff
@@ -281,7 +284,7 @@ ASTPointer<ImportDirective> Parser::parseImportDirective()
 		fatalParserError(6326_error, "Import path cannot be empty.");
 	nodeFactory.markEndPosition();
 	expectToken(Token::Semicolon);
-	return nodeFactory.createNode<ImportDirective>(path, unitAlias, move(symbolAliases));
+	return nodeFactory.createNode<ImportDirective>(path, unitAlias, unitAliasLocation, move(symbolAliases));
```
Comment (chriseth) [inline libsolidity/parsing/Parser.cpp:284]:
```suggestion
	return nodeFactory.createNode<ImportDirective>(path, unitAlias, move(unitAliasLocation), move(symbolAliases));
```

File: libsolidity/parsing/Parser.cpp [hunk 2]
```diff
@@ -234,6 +235,7 @@ ASTPointer<ImportDirective> Parser::parseImportDirective()
 		if (m_scanner->currentToken() == Token::As)
```
Comment (chriseth) [inline libsolidity/parsing/Parser.cpp:235]: Shouldn't the `else` case clear `unitAliasLocation`? This case is `import "abc.sol"`. There is no name being declared for that module, so there should be no location.

File: libsolidity/ast/AST.h [hunk 3]
```diff
@@ -244,12 +244,14 @@ class Declaration: public ASTNode, public Scopable
...
 		ASTPointer<ASTString> _name,
+		SourceLocation _nameLocation,
 		Visibility _visibility = Visibility::Default
 	):
-		ASTNode(_id, _location), m_name(std::move(_name)), m_visibility(_visibility) {}
+		ASTNode(_id, _location), m_name(std::move(_name)), m_nameLocation(std::move(_nameLocation)), m_visibility(_visibility) {}
 
 	/// @returns the declared name.
 	ASTString const& name() const { return *m_name; }
+	SourceLocation const& nameLocation() const noexcept { return m_nameLocation; }
```
Comment (chriseth) [inline libsolidity/ast/AST.h:254]: Please clarify: The location of the declared name itself or empty location if not available or unknown.

File: libsolidity/parsing/Parser.cpp (hunk 2 above)
Comment (chriseth) [reply libsolidity/parsing/Parser.cpp:235]: Actually setting it to an empty location at its definition would probably fix this.

File: libsolidity/parsing/Parser.cpp [hunk 4]
```diff
@@ -326,6 +330,7 @@ ASTPointer<ContractDefinition> Parser::parseContractDefinition()
 		documentation = parseStructuredDocumentation();
 		contractKind = parseContractKind();
 		name = expectIdentifierToken();
+		nameLocation = currentLocation();
```
Comment (chriseth) [inline libsolidity/parsing/Parser.cpp]:
```suggestion
		nameLocation = currentLocation();
		name = expectIdentifierToken();
```

File: libsolidity/parsing/Parser.cpp [hunk 5]
```diff
@@ -599,13 +607,17 @@ ASTPointer<ASTNode> Parser::parseFunctionDefinition(bool _freeFunction)
...
+			nameLocation = currentLocation();
 			name = expectIdentifierToken();
+		}
 	}
 	else
 	{
 		solAssert(kind == Token::Constructor || kind == Token::Fallback || kind == Token::Receive, "");
 		m_scanner->next();
 		name = make_shared<ASTString>();
+		nameLocation = currentLocation();
```
Comment (chriseth) [inline libsolidity/parsing/Parser.cpp]: I think this has to be empty here - the only other option would be to be around the keyword (i.e. above the call to `next`), but I'm not sure that will work.

File: libsolidity/parsing/Parser.cpp [hunk 6]
```diff
@@ -641,6 +654,7 @@ ASTPointer<StructDefinition> Parser::parseStructDefinition()
 	ASTNodeFactory nodeFactory(*this);
 	expectToken(Token::Struct);
 	ASTPointer<ASTString> name = expectIdentifierToken();
+	SourceLocation nameLocation = currentLocation();
```
Comment (chriseth) [inline libsolidity/parsing/Parser.cpp]:
```suggestion
	SourceLocation nameLocation = currentLocation();
	ASTPointer<ASTString> name = expectIdentifierToken();
```

File: libsolidity/parsing/Parser.cpp [hunk 7]
```diff
@@ -667,6 +681,8 @@ ASTPointer<EnumDefinition> Parser::parseEnumDefinition()
 	ASTNodeFactory nodeFactory(*this);
 	expectToken(Token::Enum);
 	ASTPointer<ASTString> name = expectIdentifierToken();
```
Comment (chriseth) [inline libsolidity/parsing/Parser.cpp]: Why can't you use `currentLocation` here?

File: libsolidity/parsing/Parser.cpp [hunk 8]
```diff
@@ -831,6 +850,7 @@ ASTPointer<ModifierDefinition> Parser::parseModifierDefinition()
 
 	expectToken(Token::Modifier);
 	ASTPointer<ASTString> name(expectIdentifierToken());
+	SourceLocation nameLocation = currentLocation();
```
Comment (chriseth) [inline libsolidity/parsing/Parser.cpp]:
```suggestion
	SourceLocation nameLocation = currentLocation();
	ASTPointer<ASTString> name(expectIdentifierToken());
```

Comment (chriseth) [reply libsolidity/parsing/Parser.cpp]: Maybe it would make sense to turn this into a helper?

File: libsolidity/parsing/Parser.cpp [hunk 9]
```diff
@@ -875,7 +895,7 @@ ASTPointer<ModifierDefinition> Parser::parseModifierDefinition()
 	else
 		m_scanner->next(); // just consume the ';'
 
-	return nodeFactory.createNode<ModifierDefinition>(name, documentation, parameters, isVirtual, overrides, block);
+	return nodeFactory.createNode<ModifierDefinition>(name, nameLocation, documentation, parameters, isVirtual, overrides, block);
```
Comment (chriseth) [inline libsolidity/parsing/Parser.cpp:885]: move - several other places, too.

File: libsolidity/parsing/Parser.cpp [hunk 10]
```diff
@@ -886,6 +906,7 @@ ASTPointer<EventDefinition> Parser::parseEventDefinition()
 
 	expectToken(Token::Event);
 	ASTPointer<ASTString> name(expectIdentifierToken());
+	SourceLocation nameLocation = currentLocation();
```
Comment (chriseth) [inline libsolidity/parsing/Parser.cpp]:
```suggestion
	SourceLocation nameLocation = currentLocation();
	ASTPointer<ASTString> name(expectIdentifierToken());
```

**Comment by chriseth** [comment]
Shouldn't we export (and re-import) this in the JSON-AST? I would expect many people to be happy to have the name location.

File: libsolidity/parsing/Parser.cpp (hunk 2 above)
Comment (christianparpart) [reply libsolidity/parsing/Parser.cpp:235]: yes, thx.

File: libsolidity/parsing/Parser.cpp (hunk 4 above)
Comment (christianparpart) [reply libsolidity/parsing/Parser.cpp]: I missed this. thanks

File: libsolidity/parsing/Parser.cpp (hunk 7 above)
Comment (christianparpart) [reply libsolidity/parsing/Parser.cpp]: done. thanks.

File: libsolidity/parsing/Parser.cpp (hunk 8 above)
Comment (christianparpart) [reply libsolidity/parsing/Parser.cpp]: I agree. Done.

**Comment by christianparpart** [comment]
> Shouldn't we export (and re-import) this in the JSON-AST? I would expect many people to be happy to have the name location.

absolutely. I'll be adding that tomorrow morning (late read, sorry).

**Comment by chriseth** [comment]
Please add a changelog entry.

File: libsolidity/ast/AST.h [hunk 11]
```diff
@@ -345,9 +348,10 @@ class ImportDirective: public Declaration
 		SourceLocation const& _location,
 		ASTPointer<ASTString> _path,
 		ASTPointer<ASTString> const& _unitAlias,
+		SourceLocation _unitAliasLocation,
 		SymbolAliasList _symbolAliases
 	):
-		Declaration(_id, _location, _unitAlias),
+		Declaration(_id, _location, _unitAlias, std::move(_unitAliasLocation)),
```
Comment (chriseth) [inline libsolidity/ast/AST.h:354]: If there is one newline after a `,`, there has to be one for each of them.

File: libsolidity/ast/ASTJsonConverter.cpp [hunk 12]
```diff
@@ -117,6 +117,9 @@ size_t ASTJsonConverter::sourceIndexFromLocation(SourceLocation const& _location
 
 string ASTJsonConverter::sourceLocationToString(SourceLocation const& _location) const
 {
+	if (_location == SourceLocation{})
```
Comment (chriseth) [inline libsolidity/ast/ASTJsonConverter.cpp]: How did this work before?

File: libsolidity/ast/ASTJsonConverter.cpp [hunk 13]
```diff
@@ -243,13 +246,16 @@ bool ASTJsonConverter::visit(ImportDirective const& _node)
 	addIfSet(attributes, "absolutePath", _node.annotation().absolutePath);
 
 	attributes.emplace_back("unitAlias", _node.name());
+	attributes.emplace_back(make_pair("nameLocation", Json::Value(sourceLocationToString(_node.nameLocation()))));
```
Comment (chriseth) [inline libsolidity/ast/ASTJsonConverter.cpp]: Do you need `make_pair`?

File: libsolidity/parsing/Parser.cpp [hunk 14]
```diff
@@ -599,13 +606,16 @@ ASTPointer<ASTNode> Parser::parseFunctionDefinition(bool _freeFunction)
 			m_scanner->next();
 		}
 		else
-			name = expectIdentifierToken();
+		{
+			tie(name, nameLocation) = expectIdentifierWithLocation();
+		}
```
Comment (chriseth) [inline libsolidity/parsing/Parser.cpp]:
```suggestion
			tie(name, nameLocation) = expectIdentifierWithLocation();
```

File: libsolidity/parsing/Parser.cpp [hunk 15]
```diff
@@ -264,6 +266,7 @@ ASTPointer<ImportDirective> Parser::parseImportDirective()
 		{
 			m_scanner->next();
 			expectToken(Token::As);
+			unitAliasLocation = currentLocation();
```
Comment (chriseth) [inline libsolidity/parsing/Parser.cpp]: Can't you use `expectIdentifierWithLocation` here?

Comment (chriseth) [reply libsolidity/parsing/Parser.cpp]: (also above)

File: libsolidity/parsing/Parser.cpp [hunk 16]
```diff
@@ -599,13 +606,16 @@ ASTPointer<ASTNode> Parser::parseFunctionDefinition(bool _freeFunction)
...
+		{
+			tie(name, nameLocation) = expectIdentifierWithLocation();
+		}
 	}
 	else
 	{
 		solAssert(kind == Token::Constructor || kind == Token::Fallback || kind == Token::Receive, "");
 		m_scanner->next();
 		name = make_shared<ASTString>();
+		nameLocation = currentLocation();
```
Comment (chriseth) [inline libsolidity/parsing/Parser.cpp]: is this empty?

File: libsolidity/parsing/Parser.cpp [hunk 17]
```diff
@@ -641,6 +652,7 @@ ASTPointer<StructDefinition> Parser::parseStructDefinition()
 	ASTNodeFactory nodeFactory(*this);
 	expectToken(Token::Struct);
 	ASTPointer<ASTString> name = expectIdentifierToken();
```
Comment (chriseth) [inline libsolidity/parsing/Parser.cpp]: Again, expectIdentifierWithLocation

File: libsolidity/parsing/Parser.cpp [hunk 18]
```diff
@@ -875,7 +892,15 @@ ASTPointer<ModifierDefinition> Parser::parseModifierDefinition()
...
-	return nodeFactory.createNode<ModifierDefinition>(name, documentation, parameters, isVirtual, overrides, block);
+	return nodeFactory.createNode<ModifierDefinition>(name, nameLocation, documentation, parameters, isVirtual, overrides, block);
+}
+
+pair<ASTPointer<ASTString>, SourceLocation> Parser::expectIdentifierWithLocation()
+{
+	SourceLocation nameLocation = currentLocation();
+	ASTPointer<ASTString> name = expectIdentifierToken();
+
+	return {name, nameLocation};
```
Comment (chriseth) [inline libsolidity/parsing/Parser.cpp]: move?

File: test/cmdlineTests/recovery_ast_constructor/output [hunk 19]
```diff
@@ -139,6 +140,7 @@ JSON AST (compact format):
                 "id": 12,
                 "mutability": "mutable",
                 "name": "",
+                "nameLocation": "-1:-1:18446744073709551615",
```
Comment (chriseth) [inline test/cmdlineTests/recovery_ast_constructor/output]: The last number does not seem correct...

File: libsolidity/ast/ASTJsonConverter.cpp (hunk 12 above)
Comment (christianparpart) [reply libsolidity/ast/ASTJsonConverter.cpp]: It didn't. 

That yielded into source indices with value `numeric_limits<size_t>::max()` and that caused a throw in `std::stoi()` on the decoding side, so I tried to properly handle in the decoding side but realized that it looks like an unclean solution and the `ASTImportTest.sh` kept failing. So I decided for this solution (I think is the shortest & cleaned?)

**Comment by christianparpart** [comment]
I just realize we have quite a few (not to say many) places where we pair a name with a location while parsing. I wonder why this one isn't then an `Identifier` AST node, I'd love that to happen, also as it's easier to traverse through the information then (at least it looks more in line with the rest). I think I could make that as part of a split-PR (not in here, to get that one merged).

File: test/cmdlineTests/recovery_ast_constructor/output (hunk 19 above)
Comment (christianparpart) [reply test/cmdlineTests/recovery_ast_constructor/output]: That's the one I was referring to in the call. It's `numeric_limits<size_t>::max()`.

File: libsolidity/parsing/Parser.cpp (hunk 16 above)
Comment (christianparpart) [reply libsolidity/parsing/Parser.cpp]: yes, just like name is empty. I dropped both lines to better designate the "empty"-state.

**Comment by chriseth** [comment]
Test failures.

**Comment by chriseth** [comment]
Needs a changelog entry.

File: liblangutil/SourceLocation.cpp [hunk 20]
```diff
@@ -34,17 +34,20 @@ SourceLocation const parseSourceLocation(std::string const& _input, std::string
 
 	boost::algorithm::split(pos, _input, boost::is_any_of(":"));
 
+	auto const sourceIndex = stoi(pos[Index]);
```
Comment (chriseth) [inline liblangutil/SourceLocation.cpp:38]: Please check `pos.size() == 3` before accessing `pos[Index]`

File: libsolidity/ast/ASTJsonConverter.cpp [hunk 21]
```diff
@@ -107,21 +107,21 @@ void ASTJsonConverter::setJsonNode(
...
 
 string ASTJsonConverter::sourceLocationToString(SourceLocation const& _location) const
 {
-	size_t sourceIndex = sourceIndexFromLocation(_location);
+	optional<size_t> sourceIndexOpt = sourceIndexFromLocation(_location);
 	int length = -1;
 	if (_location.start >= 0 && _location.end >= 0)
 		length = _location.end - _location.start;
-	return to_string(_location.start) + ":" + to_string(length) + ":" + to_string(sourceIndex);
+	return to_string(_location.start) + ":" + to_string(length) + ":" + (sourceIndexOpt.has_value() ? to_string(sourceIndexOpt.value()) : "-1");
```
Comment (chriseth) [inline libsolidity/ast/ASTJsonConverter.cpp:124]:
```suggestion
	return to_string(_location.start) + ":" + to_string(length) + ":" + sourceIndexOpt.value_or("-1");
```

File: libsolidity/ast/ASTJsonImporter.cpp [hunk 22]
```diff
@@ -99,6 +99,17 @@ SourceLocation const ASTJsonImporter::createSourceLocation(Json::Value const& _n
 	return solidity::langutil::parseSourceLocation(_node["src"].asString(), m_currentSourceName, m_sourceLocations.size());
 }
 
+SourceLocation ASTJsonImporter::createNameSourceLocation(Json::Value const& _node)
+{
+	astAssert(member(_node, "nameLocation").isString(), "'nameLocation' must be a string");
+
+	string const nameLocationValue = _node["nameLocation"].asString();
+	if (nameLocationValue.empty())
```
Comment (chriseth) [inline libsolidity/ast/ASTJsonImporter.cpp]: Can this happen?

File: libsolidity/parsing/Parser.cpp [hunk 23]
```diff
@@ -226,16 +226,14 @@ ASTPointer<ImportDirective> Parser::parseImportDirective()
 	expectToken(Token::Import);
 	ASTPointer<ASTString> path;
 	ASTPointer<ASTString> unitAlias = make_shared<string>();
+	SourceLocation unitAliasLocation = currentLocation();
```
Comment (chriseth) [inline libsolidity/parsing/Parser.cpp]: Shouldn't this get the default value instead?

**Review by chriseth — CHANGES_REQUESTED** [review]

File: libsolidity/parsing/Parser.cpp [hunk 24]
```diff
@@ -264,6 +261,7 @@ ASTPointer<ImportDirective> Parser::parseImportDirective()
 		{
 			m_scanner->next();
 			expectToken(Token::As);
+			unitAliasLocation = currentLocation();
```
Comment (chriseth) [inline libsolidity/parsing/Parser.cpp]: use `expectIdentifierWithLocation`

File: libsolidity/parsing/Parser.cpp [hunk 25]
```diff
@@ -586,6 +587,7 @@ ASTPointer<ASTNode> Parser::parseFunctionDefinition(bool _freeFunction)
 				{Token::Fallback, "fallback function"},
 				{Token::Receive, "receive function"},
 			}.at(m_scanner->currentToken());
+			nameLocation = currentLocation();
```
Comment (chriseth) [inline libsolidity/parsing/Parser.cpp:592]: expectIdentifierTokenWithLocation

File: libsolidity/parsing/Parser.cpp [hunk 26]
```diff
@@ -831,6 +838,7 @@ ASTPointer<ModifierDefinition> Parser::parseModifierDefinition()
 
 	expectToken(Token::Modifier);
 	ASTPointer<ASTString> name(expectIdentifierToken());
+	SourceLocation nameLocation = currentLocation();
```
Comment (chriseth) [inline libsolidity/parsing/Parser.cpp]: please use the helper

File: libyul/AsmJsonConverter.cpp [hunk 27]
```diff
@@ -187,7 +187,7 @@ Json::Value AsmJsonConverter::createAstNode(langutil::SourceLocation const& _loc
 	int length = -1;
 	if (_location.start >= 0 && _location.end >= 0)
 		length = _location.end - _location.start;
-	ret["src"] = to_string(_location.start) + ":" + to_string(length) + ":" + m_sourceIndex;
+	ret["src"] = to_string(_location.start) + ":" + to_string(length) + ":" + (m_sourceIndex.has_value() ? to_string(m_sourceIndex.value()) : "-1");
```
Comment (chriseth) [inline libyul/AsmJsonConverter.cpp:190]: `value_or`

File: libyul/AsmJsonConverter.h [hunk 28]
```diff
@@ -40,7 +41,7 @@ class AsmJsonConverter: public boost::static_visitor<Json::Value>
 public:
 	/// Create a converter to JSON for any block of inline assembly
 	/// @a _sourceIndex to be used to abbreviate source name in the source locations
-	explicit AsmJsonConverter(size_t _sourceIndex): m_sourceIndex(std::to_string(_sourceIndex)) {}
+	explicit AsmJsonConverter(std::optional<size_t> _sourceIndex): m_sourceIndex(_sourceIndex) {}
```
Comment (chriseth) [inline libyul/AsmJsonConverter.h:44]: Doesn't this need changes in the cpp file?

File: libsolidity/parsing/Parser.cpp (hunk 23 above)
Comment (christianparpart) [reply libsolidity/parsing/Parser.cpp]: yes.

File: libsolidity/parsing/Parser.cpp [hunk 29]
```diff
@@ -245,13 +243,12 @@ ASTPointer<ImportDirective> Parser::parseImportDirective()
 			while (true)
 			{
 				ASTPointer<ASTString> alias;
-				SourceLocation aliasLocation = currentLocation();
```
Comment (chriseth) [inline libsolidity/parsing/Parser.cpp]: This should not be changed.

File: libyul/AsmJsonConverter.cpp (hunk 27 above)
Comment (chriseth) [reply libyul/AsmJsonConverter.cpp:190]: It's actually shorter like this.

**Comment by chriseth** [comment]
I hope that soltest will fail. If it does not, our test setup is broken.

**Comment by chriseth** [comment]
Phew. Everything failed as expected. Will update expectations.

**Review by chriseth — APPROVED** [review]

**Comment by Marenz** [comment]
Didn't this introduce a backwards-incompatible change and thus needs a major version change?

**Comment by christianparpart** [comment]
AFAIR it only adds (nameLocation) to the C++ AST but doesn't change nor remove. So this is backwards incompatible.

**Pull request merged**